### DISCUSSION TITLE
Colorize and customize math symbols

### DISCUSSION THREAD

**Discussion by Easyoakland** [typst/typst#2205] [Q&A]
How would one edit the content that is output from a math block?
E.g. A function that colorizes parenthesis based upon their depth.

I don't see a way to do this while preserving the rendered output of the math block.

The closest I see is:
1. Manually converting a `content` block to a `str`
2. Iterating over each character adding `text( #color, INNER_SYMBOL)`
3. Recursively replacing the `INNER_SYMBOL` as new characters are iterated over
4. Using `eval` on the result which might look like `$text(#red, {text(#green, {text(#black, 2/3)})})$`
 
But this has some issues, seems unnecessarily complicated, and just generally appears improper.

The goal would be to make a function that does this (call it `colorize_paren`) and then use it in a show rule like `#show math.equation: it => colorize_paren(it)` or `#show math.equation: colorize_paren`

**Comment by T0mstone** [comment]
Here's a solution that uses `state` and `locate`.
```typst
#let get-color(i) = (red, green, blue).at(calc.rem(i, 3)) // or whatever you want

#let depth = state("depth", 0)
#show math.equation: it => {
	show "(": it => locate(loc => text(get-color(depth.at(loc)), it)) + depth.update(s => s + 1)
	show ")": it => depth.update(s => s - 1) + locate(loc => text(get-color(depth.at(loc)), it))
	it
}

$ (((((()))))) $
```

![a](https://github.com/typst/typst/assets/39707032/75fc5090-7f47-4625-a330-949fea18df8a)

**Reply by Easyoakland** [reply]
That's closer and a much nicer, but doesn't work with things like `((((((4/3))))))` because the rightward parenthesis doesn't resize correctly.

**Reply by Easyoakland** [reply]
There is now an open issue #2357 closely related to this.
